README.rst
..
    This file is part of Invenio.
    Copyright (C) 2015, 2016 CERN.

    Invenio is free software; you can redistribute it
    and/or modify it under the terms of the GNU General Public License as
    published by the Free Software Foundation; either version 2 of the
    License, or (at your option) any later version.

    Invenio is distributed in the hope that it will be
    useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Invenio; if not, write to the
    Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
    MA 02111-1307, USA.

    In applying this license, CERN does not
    waive the privileges and immunities granted to it by virtue of its status
    as an Intergovernmental Organization or submit itself to any jurisdiction.

======================
 Invenio-OAIHarvester
======================

.. image:: https://img.shields.io/travis/inveniosoftware/invenio-oaiharvester.svg
        :target: https://travis-ci.org/inveniosoftware/invenio-oaiharvester

.. image:: https://img.shields.io/coveralls/inveniosoftware/invenio-oaiharvester.svg
        :target: https://coveralls.io/r/inveniosoftware/invenio-oaiharvester

.. image:: https://img.shields.io/github/tag/inveniosoftware/invenio-oaiharvester.svg
        :target: https://github.com/inveniosoftware/invenio-oaiharvester/releases

.. image:: https://img.shields.io/pypi/dm/invenio-oaiharvester.svg
        :target: https://pypi.python.org/pypi/invenio-oaiharvester

.. image:: https://img.shields.io/github/license/inveniosoftware/invenio-oaiharvester.svg
        :target: https://github.com/inveniosoftware/invenio-oaiharvester/blob/master/LICENSE


Invenio module for OAI-PMH metadata harvesting between repositories.

* Free software: GPLv2 license
* Documentation: https://invenio-oaiharvester.readthedocs.org.

*This is an experimental development preview release.*

Features
========

This module allows you to easily harvest OAI-PMH repositories, thanks to the `Sickle`_ module, and via signals
you can hook the output into your application, or simply to files.

You keep configurations of your OAI-PMH sources via SQLAlchemy models and run or schedule immediate harvesting jobs
via command-line or regularly via `Celery beat`_.

.. _Celery beat: http://celery.readthedocs.org/en/latest/userguide/periodic-tasks.html
.. _Sickle: http://sickle.readthedocs.org/en/latest/

Harvesting is simple
====================

.. code-block:: shell

    inveniomanage oaiharvester get -u http://export.arxiv.org/oai2 -i oai:arXiv.org:1507.07286 > my_record.xml


This will harvest the repository for a specific record and print the records to stdout - which in this case will save it to a file called ``my_record.xml``.

If you want to have your harvested records saved in a directory automatically, its easy:

.. code-block:: shell

    inveniomanage oaiharvester get -u http://export.arxiv.org/oai2 -i oai:arXiv.org:1507.07286 -d /tmp


Note the directory ``-d`` parameter that specifies a directory to save harvested XML files.


Integration with your application
=================================

If you want to integrate ``invenio-oaiharvester`` into your application, you should hook into
the signals sent by the harvester upon completed harvesting.

See ``invenio_oaiharvester.signals:oaiharvest_finished``.

Check also the defined Celery tasks under ``invenio_oaiharvester.tasks``.


Managing OAI-PMH sources
========================

If you want to store configuration for an OAI repository, you can use the
SQLAlchemy model ``invenio_oaiharvester.models:OaiHARVEST``.

This is useful if you regularly need to query a server.

Here you can add information about the server URL, metadataPrefix to use etc.
This information is also available when scheduling and running tasks:

.. code-block:: shell

    inveniomanage oaiharvester get -n somerepo -i oai:example.org:1234

Here we are using the `-n, --name` parameter to specify which configured
OAI-PMH source to query, using the ``name`` property.


API
===

If you need to schedule or run harvests via Python, you can use our API:

.. code-block:: python

    from invenio_oaiharvester.api import get_records

    request, records = get_records(identifiers=["oai:arXiv.org:1207.7214"],
                                   url="http://export.arxiv.org/oai2")
    for record in records:
        print rec.raw


CHANGES.rst
..
    This file is part of Invenio.
    Copyright (C) 2015 CERN.

    Invenio is free software; you can redistribute it
    and/or modify it under the terms of the GNU General Public License as
    published by the Free Software Foundation; either version 2 of the
    License, or (at your option) any later version.

    Invenio is distributed in the hope that it will be
    useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Invenio; if not, write to the
    Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
    MA 02111-1307, USA.

    In applying this license, CERN does not
    waive the privileges and immunities granted to it by virtue of its status
    as an Intergovernmental Organization or submit itself to any jurisdiction.

Changes
=======

Version 1.0.0a1 (released 2016-06-13)

Incompatible changes
~~~~~~~~~~~~~~~~~~~~

- Brings a major rewrite based on Invenio v3.0.0 release series.


Version 0.1.1 (released 2015-08-25)

- Adds missing `invenio_upgrader` dependency and amends past upgrade
  recipes following its separation into standalone package.

- Fixes imports due to separation of invenio-workflows. (#9)

Version 0.1.0 (released 2015-08-14)

- Initial public release


invenio_oaiharvester/manage.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2015, 2016 CERN.
#
# Invenio is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

"""CLI tool to harvest records from an OAI-PMH repository."""

from __future__ import absolute_import, print_function, unicode_literals

from invenio_ext.script import Manager

from .errors import IdentifiersOrDates
from .tasks import get_specific_records, list_records_from_dates
from .utils import (
    write_to_dir,
    print_to_stdout,
    print_total_records,
    print_files_created,
)

manager = Manager(description=__doc__)


@manager.option('-m', '--metadataprefix', dest='metadata_prefix', default=None,
                help="The prefix for the metadata return (e.g. 'oai_dc')")
@manager.option('-n', '--name', dest='name', default=None,
                help="The name of the OaiHARVEST object that we want to use to create the endpoint.")
@manager.option('-s', '--setSpec', dest='setSpec', default=None,
                help="The 'set' criteria for the harvesting (optional).")
@manager.option('-i', '--identifiers', dest='identifiers', default=None,
                help="A list of unique identifiers for records to be harvested.")
@manager.option('-f', '--from', dest='from_date', default=None,
                help="The lower bound date for the harvesting (optional).")
@manager.option('-t', '--to', dest='until_date', default=None,
                help="The upper bound date for the harvesting (optional).")
@manager.option('-u', '--url', dest='url', default=None,
                help="The upper bound date for the harvesting (optional).")
@manager.option('-d', '--dir', dest='directory', default=None,
                help="The directory that we want to send the harvesting results.")
@manager.option('-a', '--args', dest='arguments', default=[], action="append",
                help="Arguments to harvesting task, in the form `-a arg1=val1`.")
@manager.option('-q', '--quiet', dest='quiet', action="store_true", default=False,
                help="Surpress output.")
@manager.option('-k', '--enqueue', dest='enqueue', action="store_true", default=False,
                help="Enqueue harvesting and return immediately.")
@manager.option('-p', '--no-signals', dest='signals', action="store_false", default=True,
                help="No signals sent with OAI-PMH harvesting results.")
def harvest(metadata_prefix, name, setSpec, identifiers, from_date,
            until_date, url, directory, arguments, quiet, enqueue, signals):
    """Harvest records from an OAI repository."""
    arguments = dict(x.split('=', 1) for x in arguments)
    if identifiers is None:
        # If no identifiers are provided, a harvest is scheduled:
        # - url / name is used for the endpoint
        # - from_date / lastrun is used for the dates (until_date optionally if from_date is used)
        params = (metadata_prefix, from_date, until_date, url,
                  name, setSpec, directory, signals)
        if enqueue:
            job = list_records_from_dates.delay(*params, **arguments)
            print("Scheduled job {0}".format(job.id))
        else:
            records = list_records_from_dates(*params, **arguments)
    else:
        if (from_date is not None) or (until_date is not None):
            raise IdentifiersOrDates("Identifiers cannot be used in combination with dates.")

        # If identifiers are provided, we schedule an immediate run using them.
        params = (identifiers, metadata_prefix, url,
                  name, directory, signals)
        if enqueue:
            job = get_specific_records.delay(*params, **arguments)
            print("Scheduled job {0}".format(job.id))
        else:
            records = get_specific_records(*params, **arguments)
    if directory:
        files_created, total = write_to_dir(records, directory)
        print_files_created(files_created)
        print_total_records(total)
    elif not quiet:
        total = print_to_stdout(records)
        print_total_records(total)


def main():
    """Run manager."""
    from invenio_base.factory import create_app
    app = create_app()
    manager.app = app
    manager.run()

if __name__ == '__main__':
    main()
